### ISSUE TITLE
Minification takes >40min

### ISSUE THREAD

**Issue by pisti2010** [meteor/meteor#10480]
For a while now we have been experiencing extremely long minification times for our applicatio. It can take from 30min to 1hr to be able to deploy a build to production.

Meteor version: 1.8.0.1
OS: Linux Ubuntu

I have done an extensive search through the various issues found here and elsewhere (Stackoverflow) to find a solution. I have tried other minifiers, "hacking" the minifier to find if it is failing and using the babili minifier, and other similar procedures without success. Hence I am creating this issue to see if I can get some help to fix this issue that is making our release process a pain.

Packages:
meteor-base@1.4.0             
mobile-experience@1.0.5   
mongo@1.6.0                   
blaze-html-templates@1.0.4 
reactive-var@1.0.11            
tracker@1.2.0                 
standard-minifier-css@1.5.2  
standard-minifier-js@2.4.0   
es5-shim@4.8.0                
ecmascript@0.12.3           
shell-server@0.4.0           

accounts-ui@1.3.1
accounts-google@1.3.2
google-config-ui@1.0.1
gadicc:blaze-react-component
react-meteor-data
alanning:roles
aldeed:collection2-core
percolate:migrations
service-configuration@1.0.11
pbastowski:body-parser
qualia:profile
mixmax:smart-disconnect
underscore@1.0.10
montiapm:profiler
montiapm:agent
email@1.2.3
meteorhacks:ssr
deanius:promise
ros:publish-counts
reywood:publish-composite
meteortesting:mocha
xolvio:cleaner
mdg:validation-error

 
Profile of the build process:

$ METEOR_PROFILE=1 meteor build ../out-test
| (#1) Profiling: meteor build
...
|
| Top leaves:
| other Target#minifyJs................................1,944,835 ms (2)
| files.createTarball web.tar.gz.........................478,989 ms (1)
| files.readFile..........................................68,233 ms (112140)
| files.rm_recursive......................................35,713 ms (5)
| files.writeFile.........................................33,882 ms (98276)
| other optimistic lstatOrNull............................29,491 ms (171851)
| other Builder#copyNodeModulesDirectory..................28,215 ms (33)
| other optimistic lstat..................................27,361 ms (118318)
| other optimistic statOrNull.............................21,098 ms (247124)
| other optimistic readFile...............................16,060 ms (119388)
| other ImportScanner#_findImportedModuleIdentifiers......15,251 ms (15425)
| files.stat..............................................12,182 ms (152393)
| files.lstat.............................................11,040 ms (111973)
| files.readdir............................................6,812 ms (28052)
| other Builder#_ensureDirectory...........................5,055 ms (8319)
| other ImportScanner#_resolve.............................3,248 ms (58812)
| files.mkdir..............................................2,865 ms (8377)
| other optimistic readJsonOrNull..........................2,689 ms (5788)
| other ImportScanner#scanImports for the app..............2,199 ms (3)
| other files.withCache....................................2,110 ms (7)
| Babel.compile............................................1,868 ms (754)
| other optimistic readdir.................................1,866 ms (8363)
| CssTools.minifyCss.......................................1,388 ms (4)
| other bundler writeFile..................................1,384 ms (20)
| other safeWatcher.watch..................................1,156 ms (7751)
| other PackageSourceBatch#_linkJS...........................971 ms (296)
| other ImportScanner#_readFile..............................921 ms (15192)
| other ImportScanner#_realPath..............................849 ms (16203)
| other JsImage#write........................................731 ms (1)
| sha512.....................................................573 ms (228)
| sha1.......................................................530 ms (12487)
| files.realpath.............................................507 ms (10193)
| other ImportScanner#scanImports for base64.................464 ms (3)
| other optimistic hashOrNull................................438 ms (16924)
| ImportScanner#_getAbsModuleId..............................387 ms (17292)
| require("./security.js")...................................300 ms (4)
| other meteorNpm.isPortable.................................260 ms (2904)
| other Target#make..........................................227 ms (3)
| other ImportScanner#scanImports for montiapm:agent.........196 ms (3)
| other Target#_runCompilerPlugins...........................192 ms (3)
| other PackageSourceBatch.computeJsOutputFilesMap...........171 ms (3)
| other Npm.require("terser")................................164 ms (1)
| wrapped.fs.readFileSync....................................161 ms (628)
| files.read.................................................131 ms (1428)
| other plugin templating-compiler...........................126 ms (2)
| plugin ecmascript..........................................101 ms (3)
|
| 
(#1) Total: 2,770,654 ms (meteor build)

**Comment by StorytellerCZ** [comment]
Can you please also provide your npm dependencies?

**Comment by pisti2010** [comment]
NPM Dependencies:
    "@babel/runtime": "^7.2.0",
    "@material-ui/core": "^3.6.1",
    "@material-ui/icons": "^3.0.1",
    "@splitsoftware/splitio": "^10.6.0",
    "arraybuffer-base64": "^1.0.0",
    "aws-sdk": "^2.370.0",
    "axios": "^0.18.0",
    "babel-runtime": "^6.26.0",
    "body-parser": "^1.18.3",
    "bufferutil": "^4.0.1",
    "classnames": "^2.2.6",
    "createjs-module": "^0.8.3",
    "cron": "^1.5.1",
    "gl-vec2": "^1.3.0",
    "handlebars": "^4.0.12",
    "history": "^4.7.2",
    "influx": "^5.0.7",
    "interactjs": "^1.3.4",
    "mapbox-gl": "^0.53.0-beta.1",
    "meteor-node-stubs": "~0.2.4",
    "moment": "^2.22.2",
    "mqtt": "^2.18.8",
    "pako": "^1.0.7",
    "password-generator": "^2.2.0",
    "plotly.js": "^1.44.4",
    "prop-types": "^15.6.2",
    "protobufjs": "^6.8.8",
    "query-string": "^5.1.1",
    "radium": "^0.25.1",
    "react": "~16.8.1",
    "react-ad-block-detect": "^1.0.1",
    "react-addons-pure-render-mixin": "^15.6.0",
    "react-calendar-timeline": "^0.18.2",
    "react-color": "^2.17.0",
    "react-copy-to-clipboard": "^5.0.1",
    "react-d3-speedometer": "^0.3.3",
    "react-dom": "~16.8.1",
    "react-dropzone": "^8.0.2",
    "react-ga": "^2.5.6",
    "react-google-button": "^0.4.0",
    "react-hotkeys": "^1.1.4",
    "react-load-script": "0.0.6",
    "react-map-gl": "^4.0.10",
    "react-map-gl-geocoder": "^1.6.3",
    "react-nipple": "^1.0.1",
    "react-perf-devtool": "^3.1.8",
    "react-plotly.js": "^2.2.0",
    "react-portal": "^4.2.0",
    "react-router": "^4.3.1",
    "react-select": "^2.1.2",
    "react-sortable-hoc": "^0.8.4",
    "react-split-pane": "^0.1.85",
    "react-stripe-elements": "^2.0.1",
    "react-swipeable-views": "^0.12.18",
    "simpl-schema": "^1.5.3",
    "slack": "^11.0.1",
    "stripe": "^6.17.0",
    "typeface-roboto": "0.0.35",
    "uglify-js-es6": "^2.8.9",
    "utf-8-validate": "^5.0.1"

**Comment by KoenLav** [comment]
Did you set the NODE_TOOLS environment variable with max-old-space-size?

**Comment by pisti2010** [comment]
Yes, in the build/deployment script we use this:
export TOOL_NODE_FLAGS=--max-old-space-size=4096

**Comment by smeijer** [comment]
Is it on a virtual machine? How many CPU's are available?

**Comment by pisti2010** [comment]
Builds have been done both locally in Linux Ubuntu 16.04 not VM environments, and in Cloud VM environments. 
CPU availability does not seem to be an issue, for example in my environment I have not seen any of my 4 cores maxing out at any moment during minification, they sit at around 50% capacity. Others in my team with more powerful hardware have practically the same minification times.

I had read about the standard minifier potentially failing and a babili minifier taking its place, which is quite slow, but I believe this is outdated information. Any clues on that front? Thanks btw, any clue/question is welcomed.

**Comment by zhang2118** [comment]
Our team is also suffering from slow js minification. Not as dramatic as 30 min, but it does take much longer time to finish compared with other leaves during building.

| Top leaves:
| other Target#minifyJs..................................172,896 ms (3)
| files.createTarball INSTANT.tar.gz......................75,273 ms (1)
| files.rm_recursive......................................31,683 ms (564)
| CssTools.minifyCss......................................27,775 ms (3)
| files.readFile..........................................24,866 ms (77687)
| files.writeFile.........................................22,965 ms (76497)
| other plugin less.......................................18,283 ms (3)
| meteorNpm.runNpmCommand..................................6,991 ms (3)
| files.stat...............................................5,129 ms (131854)
| getPrelinkedFiles toStringWithSourceMap..................5,069 ms (573)
| other Builder#copyNodeModulesDirectory...................4,985 ms (48)
| other optimistic lstat...................................3,922 ms (110339)
| other optimistic lstatOrNull.............................3,847 ms (100792)
| linker File#getPrelinkedOutput...........................3,846 ms (5127)
| other optimistic statOrNull..............................3,738 ms (143702)
| linker File#computeAssignedVariables.....................2,362 ms (4251)
| other optimistic readFile................................2,333 ms (78303)
| other optimistic hashOrNull..............................2,323 ms (78807)
| other ImportScanner#_findImportedModuleIdentifiers.......2,299 ms (2125)
| files.lstat..............................................2,254 ms (86199)
| other files.withCache....................................2,045 ms (18)
| files.mkdir..............................................1,669 ms (11278)
| other Builder#_ensureDirectory...........................1,534 ms (11041)
| sha1.....................................................1,481 ms (78038)
| Babel.compile............................................1,174 ms (898)
| files.readdir............................................1,118 ms (11429)
| CssTools.stringifyCss....................................1,048 ms (2)
| composing source maps....................................1,016 ms (2)
| CssTools.rewriteCssUrls..................................1,004 ms (37)
| other ImportScanner#_resolve...............................600 ms (9614)
| other bundler writeSiteArchive.............................542 ms (1)
| other meteorNpm.isPortable.................................507 ms (30240)
| files.read.................................................506 ms (1809)
| other safeWatcher.watch....................................489 ms (10451)
| CssTools.parseCss..........................................489 ms (37)
| other optimistic readdir...................................443 ms (14226)
| other optimistic readJsonOrNull............................339 ms (8139)
| files.realpath.............................................330 ms (4201)
| other ImportScanner#scanImports for the app................324 ms (4)
| optimistic readFile........................................311 ms (76899)
| other linker.fullLink......................................299 ms (573)
| other linker Module#getPrelinkedFiles......................288 ms (573)
| other JsImage#write........................................283 ms (2)
| other ImportScanner#scanImports for base64.................228 ms (4)
| other mergeCss.............................................220 ms (3)
| wrapped.fs.readFileSync....................................212 ms (383)
| other bundler writeFile....................................210 ms (549)
| require("./security.js")...................................208 ms (4)
| sha512.....................................................208 ms (1217)
| other Target#_emitResources................................157 ms (5)
| other compileUnibuild (the app)............................143 ms (4)
| ImportScanner#_findImportedModuleIdentifiers...............131 ms (335)
| other Isopack#initFromPath.................................110 ms (117)
| other runJavaScript packages/ecmascript-runtime.js.........105 ms (3)
| other ImportScanner#_realPath..............................102 ms (3480)
| other Target#_runCompilerPlugins...........................100 ms (5)
| 
| (#1) Total: 449,396 ms (meteor bundle)

Above result is from 1.8.1-beta.17. We've also tested with beta.20, but wasn't seeing any material changes regarding build time.

**Comment by hexsprite** [comment]
would it make more sense for Meteor to minify the individual files in parallel before concatenation? I heard somewhere that smaller files minify much quicker and don't make a huge difference in terms of final size

**Comment by pisti2010** [comment]
Is that a suggestion or something that we could do at the moment as a fix @hexsprite ?

**Comment by pisti2010** [comment]
This is still a very troublesome issue for us. Anyone have any other suggestions to help us? Thanks

**Comment by hexsprite** [comment]
Here's the article I found that talks about reducing minification times and taking advantage of parallelization. I'm not sure how one would accomplish this in Meteor but it would be worth thinking.

**Comment by hexsprite** [comment]
Nevermind, I was looking at the code and it seems like Meteor does minify before concatentation already. Still there is some potential for improvement by using parallizaiton.

**Comment by zodern** [comment]
I published a beta version of [zodern:standard-minifier-js](https://atmospherejs.com/zodern/standard-minifier-js) today which adds caching. You can try it with `meteor add zodern:standard-minifier-js@4.0.0-beta.1`. In a small app, minification went from 35 to 5 seconds with an up to date cache, and rebuilds spent 2 seconds on minification.

**Comment by hexsprite** [comment]
amazing!! I will try this out soon and let you know how it goes

**Comment by hexsprite** [comment]
@zodern what do you think about the idea of minifying files across multiple cores? You know the architecture of the minification system better than most. Is it practical?

**Comment by hexsprite** [comment]
wow! much faster

4.0-beta.1 (1st run)

`BRANCH_NAME=master bash devops/build_meteor.sh  519.28s user 58.40s system 83% cpu 11:33.49 total`

4.0-beta.1 (2nd time)

`BRANCH_NAME=master bash devops/build_meteor.sh  165.48s user 41.26s system 79% cpu 4:20.34 total`

**Comment by hexsprite** [comment]
@zodern wow! much faster

4.0-beta.1 (1st run)

`BRANCH_NAME=master bash devops/build_meteor.sh  519.28s user 58.40s system 83% cpu 11:33.49 total`

4.0-beta.1 (2nd time)

`BRANCH_NAME=master bash devops/build_meteor.sh  165.48s user 41.26s system 79% cpu 4:20.34 total`

**Comment by zodern** [comment]
> what do you think about the idea of minifying files across multiple cores?

It is possible, and I plan to try adding it once Meteor updates to node 10 or newer. Meteor currently gives the minifier an array of files for the main bundle, though dynamically imported files are provided one at a time.

**Comment by pisti2010** [comment]
@zodern thanks for sharing that! I am sad to report did this not work for us. Almost all our minifications happen in a CI that starts a new container for the test+build process, so the caching would not work there on the current configuration.
I did try it locally in my dev env too see if the caching worked, but without success, a few runs and all minifications/builds took the usual 30 to 40min. 
How can I verify the caching is working on my local env? If I can make this work locally, I can move towards making our CI work similarly.

**Comment by zodern** [comment]
I published 4.0.0-beta.2 which adds some logging when the `METEOR_BABELMINIFIER_CACHE_DEBUG` environment variable is set to `true`.

**Comment by hexsprite** [comment]
@pisti2010 typical CI pattern is to save the cache at the end of the build to a persistent storage and restore before the next build...

**Comment by pisti2010** [comment]
I am happy to report that updating from Meteor 1.8.0.1 to 1.8.1 has reduced our minification times to consistent 12min. 
Relevant Package updates:
```
mongo@1.6.0                  ->     mongo@1.6.2
standard-minifier-css@1.5.2  ->     standard-minifier-css@1.5.3
standard-minifier-js@2.4.0   ->     standard-minifier-js@2.4.1
ecmascript@0.12.3            ->     ecmascript@0.12.4
```
This is a very good improvement for us; however, I wonder if it can be improved some more?

**Comment by smeijer** [comment]
@zodern, hijacking the issue a bit; can you add support for an `process.env` key so that we can change the `cache path`? That way we can export the cache out of a docker container to be able to restore on the next build. Maximum profit :)

**Comment by Batistleman** [comment]
Any chance this could move up the priority list for the new MDG? It would be awesome to see some of these improvements (eg. sourcemaps) being merged into the core packages.

**Comment by jankapunkt** [comment]
Is this still relevant with Meteor 2.0?

**Comment by StorytellerCZ** [comment]
Closing this since there aren't any recent responses. If this is still a problem for you please re-open or open a new issue.
